Sync IPC message cancellation is not consistent

kkinnunen-apple opened the pull request:
#### e06af74daaf6e21925cd1b73a6cf844e8623ff92
<pre>
Sync IPC message cancellation is not consistent
<a href="https://bugs.webkit.org/show_bug.cgi?id=280271">https://bugs.webkit.org/show_bug.cgi?id=280271</a>
<a href="https://rdar.apple.com/136584036">rdar://136584036</a>

Reviewed by Matt Woodrow.

IPC::Connection has a feature where unhandled sync messages are intended
to be be cancelled. This is so that the caller does not hang
indefinitively. The cancel was sync message reply with no parameters.
This would lead to inconsistent behavior: accepted reply if the reply
does not have parameters, decode error if the reply has parameters.

Fix by following steps:

Introduce a new message, CancelSyncMessageReply, to signify that the
message was cancelled.

Replaces SyncMessageDeserializationFailure with sending the
CancelSyncMessageReply. This is used in IPC JS testing API to be able
to send multiple arbitrary incorrect messages, while avoiding timeouts.

The same structure will be used for future to implement the same cancel
feature for asynchronous messages.

This is work towards fixing leaks of asynchronous message handlers,
thread-safe sendSync/waitForAsyncReplyAndDispatchImmediately and
unification of sync/async message receive implementations.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::dispatchMessageReceiverMessage):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::dispatchSyncMessage):
(IPC::errorAsString):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendSync):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::Decoder::hasSyncMessageDeserializationFailure const): Deleted.
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::syncRequestID const):
(IPC::Decoder::isAllowedWhenWaitingForUnboundedSyncReply const):
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::Encoder::setSyncMessageDeserializationFailure): Deleted.
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/HandleMessage.h:
(IPC::handleMessageSynchronous):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::didClose):
(IPC::StreamServerConnection::dispatchStreamMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
(IPC::StreamServerConnection::processStreamMessage):
(IPC::StreamServerConnection::processOutOfStreamMessage):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::protectedWorkQueue const):
(IPC::StreamServerConnection::dispatchOutOfStreamMessage): Deleted.
(IPC::StreamServerConnection::sendDeserializationErrorSyncReply): Deleted.
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::sendSyncReply):
* Source/WebKit/Platform/IPC/SyncRequestID.h: Added.
* Source/WebKit/Scripts/webkit/model.py:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::syncMessageNotStreamEncodableBoth):
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::dispatchSyncMessage):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsResultFromReplyDecoder):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::MockTestSyncMessage::name):
(TestWebKitAPI::MockTestSyncMessageWithDataReply::name):
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/284553@main">https://commits.webkit.org/284553@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9eaadbe95998ce458a6f4e23692e83d9165c77a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69817 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22570 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20975 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71934 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57018 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20826 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/73902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13904 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72883 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44851 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60213 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/69326 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41516 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17652 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19352 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/63446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17996 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75616 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14042 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17237 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14077 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60293 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63053 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15498 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11058 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4672 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45021 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46095 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47366 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45836 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
```diff
@@ -1094,6 +1095,103 @@ TEST_P(ConnectionRunLoopTest, SendLocalSyncMessageWithDataReply)
+            EXPECT_EQ(77u, gotDestination);
+        }
+        {
+            gotDestination = 0;
```
Should there be a test for the '88' message too?

kkinnunen-apple replied:
Argh. Yeah, it was moved below. Decode errors are not within the domain of accepted errors, they're IPC attacks that can be reacted to only with IPC_TESTING_API defined + ignoreInvalidMessagesForTesting. I'll remove the hunk from this test. Good, I see that you read the test mindfully..

kkinnunen-apple committed:
Sync IPC message cancellation is not consistent
https://bugs.webkit.org/show_bug.cgi?id=280271
rdar://136584036

Reviewed by Matt Woodrow.

IPC::Connection has a feature where unhandled sync messages are intended
to be be cancelled. This is so that the caller does not hang
indefinitively. The cancel was sync message reply with no parameters.
This would lead to inconsistent behavior: accepted reply if the reply
does not have parameters, decode error if the reply has parameters.

Fix by following steps:

Introduce a new message, CancelSyncMessageReply, to signify that the
message was cancelled.

Replaces SyncMessageDeserializationFailure with sending the
CancelSyncMessageReply. This is used in IPC JS testing API to be able
to send multiple arbitrary incorrect messages, while avoiding timeouts.

The same structure will be used for future to implement the same cancel
feature for asynchronous messages.

This is work towards fixing leaks of asynchronous message handlers,
thread-safe sendSync/waitForAsyncReplyAndDispatchImmediately and
unification of sync/async message receive implementations.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::dispatchMessageReceiverMessage):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::dispatchSyncMessage):
(IPC::errorAsString):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendSync):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::Decoder::hasSyncMessageDeserializationFailure const): Deleted.
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::syncRequestID const):
(IPC::Decoder::isAllowedWhenWaitingForUnboundedSyncReply const):
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::Encoder::setSyncMessageDeserializationFailure): Deleted.
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/HandleMessage.h:
(IPC::handleMessageSynchronous):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::didClose):
(IPC::StreamServerConnection::dispatchStreamMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
(IPC::StreamServerConnection::processStreamMessage):
(IPC::StreamServerConnection::processOutOfStreamMessage):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::protectedWorkQueue const):
(IPC::StreamServerConnection::dispatchOutOfStreamMessage): Deleted.
(IPC::StreamServerConnection::sendDeserializationErrorSyncReply): Deleted.
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::sendSyncReply):
* Source/WebKit/Platform/IPC/SyncRequestID.h: Added.
* Source/WebKit/Scripts/webkit/model.py:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::syncMessageNotStreamEncodableBoth):
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::dispatchSyncMessage):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsResultFromReplyDecoder):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::MockTestSyncMessage::name):
(TestWebKitAPI::MockTestSyncMessageWithDataReply::name):
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/284553@main
